tests/dec/testPolygonalCalculus.cpp

```diff
@@ -0,0 +1,216 @@
+    { 3, 2, 6, 7 } ,
+    { 2, 0, 4, 6 } ,
+    { 4, 5, 8, 9 } };
+  
+  Mesh box(positions.cbegin(), positions.cend(),
+           faces.cbegin(), faces.cend());
+  
+  PolygonalCalculus<Mesh> boxCalculus(box);
+  
+  SECTION("Construction and basic operators")
+    {
+      REQUIRE( boxCalculus.isValid() );
+      REQUIRE( boxCalculus.nbVertices() == positions.size() );
+
+      PolygonalCalculus<Mesh>::Face f = 0;
+      auto x = boxCalculus.X(f);
+      auto d = boxCalculus.D(f);
+      auto a = boxCalculus.A(f);
+      
+      //Checking X
+      PolygonalCalculus<Mesh>::Vector vec = x.row(0);
+      REQUIRE( vecToRealPoint(vec ) == positions[1]);
+      vec = x.row(1);
+      REQUIRE( vecToRealPoint(vec ) == positions[0]);
+      vec = x.row(2);
+      REQUIRE( vecToRealPoint(vec ) == positions[2]);
+      vec = x.row(3);
+      REQUIRE( vecToRealPoint(vec ) == positions[3]);
+
+      //Some D and A
+      REQUIRE( d(1,1) == -1 );
+      REQUIRE( d(0,1) == 1 );
+      REQUIRE( d(0,2) == 0 );
+
+      REQUIRE( a(1,1) == 0.5 );
+      REQUIRE( a(0,1) == 0.5 );
+      REQUIRE( a(0,2) == 0 );
+      
+      auto vectorArea = boxCalculus.vectorArea(f);
+      
+      //Without correction, this should match
+      for(auto f=0; f<6; ++f)
+        REQUIRE( boxCalculus.faceArea(f) == box.faceArea(f) );
+     
+      box.computeFaceNormalsFromPositions();
+      for(auto f=0; f<6; ++f)
+      {
+        auto cn = boxCalculus.faceNormalAsDGtalVector(f);
+        auto n = box.faceNormal(f);
+        REQUIRE(  cn == n  );
+      }
+      
+      RealPoint c = boxCalculus.centroidAsDGtalPoint(f);
+      RealPoint expected(0.5,0.5,0.0);
+      REQUIRE(c == expected);
+    }
+  
+  SECTION("Derivatives")
+  {
+    PolygonalCalculus<Mesh>::Face f = 0;
+    auto d = boxCalculus.D(f);
+    
+    auto nf = boxCalculus.faceDegree(f);
+    PolygonalCalculus<Mesh>::Vector phi(nf),expected(nf);
+    phi << 1.0, 3.0, 2.0, 6.0;
+    expected << 2,-1,4,-5;
+    auto dphi = d*phi;  // n_f x 1 matrix
+    REQUIRE(dphi == expected);
+    
+  }
+  
+  SECTION("Structural propertes")
+  {
+    PolygonalCalculus<Mesh>::Face f = 0;
+    auto nf =  boxCalculus.faceDegree(f);
+    PolygonalCalculus<Mesh>::Vector phi(nf);
+    phi << 1.0, 3.0, 2.0, 6.0;
+    
+    auto G = boxCalculus.gradient(f);
+    auto gphi = G*phi;
+    auto coG = boxCalculus.coGradient(f);
+    auto cogphi = coG*phi;
+    
+    // grad . cograd == 0
+    REQUIRE( gphi.dot(cogphi) == 0.0);
+        
+    //    Gf = Uf Df
+    REQUIRE( G == boxCalculus.sharp(f)*boxCalculus.D(f));
+    
+    //    UV = I - nn^t (lemma4)
+    PolygonalCalculus<Mesh>::Vector n = boxCalculus.faceNormal(f);
+    REQUIRE( boxCalculus.sharp(f)*boxCalculus.flat(f) == PolygonalCalculus<Mesh>::DenseMatrix::Identity(3,3) - n*n.transpose() );
+    
+    //    P^2 = P (lemma6)
+    auto P = boxCalculus.P(f);
+    REQUIRE( P*P == P);
+    
+    //    PV=0 (lemma5)
+    REQUIRE( (P*boxCalculus.flat(f)).norm() == 0.0);
```

Comment:
Question ? There is no risk that the result is almost 0 (but not 0 numerically) ?

Reply:
👍